src/Servers/Kestrel/Core/src/Internal/Http2/Http2HeaderEnumerator.cs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
+{
+    internal struct Http2HeadersEnumerator : IEnumerator<KeyValuePair<string, string>>
+    {
+        private bool _isTrailers;
+        private HttpResponseHeaders.Enumerator _headersEnumerator;
+        private HttpResponseTrailers.Enumerator _trailersEnumerator;
+        private IEnumerator<KeyValuePair<string, StringValues>> _genericEnumerator;
+        private StringValues.Enumerator _stringValuesEnumerator;
+
+        public KeyValuePair<string, string> Current { get; private set; }
+        object IEnumerator.Current => Current;
+
+        public Http2HeadersEnumerator(HttpResponseHeaders headers)
+        {
+            _headersEnumerator = headers.GetEnumerator();
+            _trailersEnumerator = default;
+            _genericEnumerator = null;
+            _isTrailers = false;
+
+            _stringValuesEnumerator = default;
+            Current = default;
+        }
+
+        public Http2HeadersEnumerator(HttpResponseTrailers trailers)
+        {
+            _headersEnumerator = default;
+            _trailersEnumerator = trailers.GetEnumerator();
+            _genericEnumerator = null;
+            _isTrailers = true;
+
+            _stringValuesEnumerator = default;
+            Current = default;
+        }
+
+        public Http2HeadersEnumerator(IDictionary<string, StringValues> headers)
+        {
+            _headersEnumerator = default;
+            _trailersEnumerator = default;
+            _genericEnumerator = headers.GetEnumerator();
+            _isTrailers = true;
```

Comment:
nit: should _isTrailers be true here? I think it doesn't matter because you check the generic enumerator before checking isTrailers each time, but would be nice to fix.